Make menu accessible

kimberlysuazo opened the pull request:
Add accessibility to Menu, IconMenu, and MenuItem.

kimberlysuazo committed:
Add esc key event listener to Menu

kimberlysuazo committed:
Add tabIndex prop to MenuItem

Commit:
Update handleEscape to return keyboard focus to menu icon on hide

Commit:
Add a helper for creating a keyboard trap
- Can be used by all components that need to restrict keyboard focus when open, with Tab, arrow keys, etc

Commit:
Add keyboard trap to Menu
-Works with down and up arrows according to aria authoring practices

Commit:
Add appropriate aria attributes based on authoring practices

Commit:
Make MenuItem's tabIndex update on Menu open/close

Commit:
Update keyboard trap to use default export

jodonnell commented on components/menu/Menu.js:
```diff
@@ -196,6 +196,12 @@ const factory = (MenuItem) => {
     }
 
+    handleEscape = (e) => {
+      if (e.key === 'Escape') {
```
unless browser support is an issue, looks good

jodonnell commented on components/menu/MenuItem.js:
```diff
@@ -46,8 +47,8 @@ const factory = (ripple) => {
         this.handleClick(event);
       }
-    }
-    l
```
how did this make it in here?

jodonnell commented on components/menu/MenuItem.js:
```diff
@@ -72,7 +74,7 @@ const factory = (ripple) => {
           onKeyDown={this.handleEnter}
           role="menuitem"
-          tabIndex={0}
+          tabIndex={tabIndex}
```
should this be tabIndex || 0 or something so it defaults to having a tabIndex of zero?

jodonnell commented on components/menu/Menu.js:
```diff
@@ -192,13 +192,14 @@ const factory = (MenuItem) => {
       if (e.key === 'Escape') {
-        this.hide();
+        const menuIcon = ReactDOM.findDOMNode(this).parentNode.firstChild;
+        this.hide(menuIcon.focus()); // keyboard focus returns to menu icon on hide
```
menuIcon.focus() is executed immediately.  you want menuIcon.focus

jodonnell commented on components/menu/Menu.js:
```diff
@@ -162,6 +166,20 @@ const factory = (MenuItem) => {
         : undefined;
     }
 
+    handleKeyboardTrap = (event) => {
```
should this also be extracted into utils?

kimberlysuazo replied:
@jodonnell I was thinking that extracting it to utils might make it less flexible b/c this way the keyboard trap can be made to cycle only in one direction. That's also why there's a bit of repetition b/c I didnt want to couple any of the functions. Also to extract would mean having to pass in the key code for different keys ('up','down', 'left', 'right', 'tab', 'shift tab') and wouldnt allow for a keyboard trap to listen to two different keys at the same time. Do you think it makes sense to try to allow for more flexibility that way or is it better to extract it?

jodonnell replied:
sounds like its too early to generalize

jodonnell commented on components/menu/Menu.js:
```diff
@@ -162,6 +166,20 @@ const factory = (MenuItem) => {
 
+    handleKeyboardTrap = (event) => {
+      const focusableMenuItems = this.menuNode.querySelectorAll('[tabindex]');
+      const firstItem = focusableMenuItems[0];
```
maybe want to check that there array has length?

jodonnell commented on components/menu/Menu.js:
```diff
@@ -162,6 +166,20 @@ const factory = (MenuItem) => {
+        nextElementKeyboardTrap(event.target, firstItem, lastItem);
+      } else if (event.which === 38) { // on up arrow keyboard event
+        previousElementKeyboardTrap(event.target, firstItem, lastItem);
+      } else if (event.key === 'Tab') {
```
likewise on event.key

jodonnell commented on components/menu/Menu.js:
```diff
@@ -162,6 +166,20 @@ const factory = (MenuItem) => {
+      const lastItem = focusableMenuItems[focusableMenuItems.length - 1];
+
+      if (event.which === 40) { // on down arrow keyboard event
+        nextElementKeyboardTrap(event.target, firstItem, lastItem);
```
i have found i almost always want to use event.currentTarget rather than event.target

jodonnell commented on components/menu/Menu.js:
```diff
@@ -5,7 +5,7 @@ import classnames from 'classnames';
 import { themr } from 'react-css-themr';
```
this commit is a bit confusing because it uses code that isn't introduced until next commit

jodonnell commented on components/utils/keyboardTrap.js:
```diff
@@ -1,19 +1,33 @@
-    firstItem.focus();
-  }
-};
+    const firstItem = focusableElements[0];
```
these two functions are pretty duplicative.  the duplication is on the harmless side though, so your call

jodonnell commented on components/menu/Menu.js:
```diff
@@ -246,7 +246,7 @@ const factory = (MenuItem) => {
 
       return (
-        <div data-react-toolbox="menu" role="menu" className={className} style={this.getRootStyle()} onKeyDown={this.handleEscape} tabIndex={0}>
+        <div data-react-toolbox="menu" role="menu" className={className} style={this.getRootStyle()} onKeyDown={this.handleEscape} tabIndex={this.state.tabIndex}>
```
instead of keeping tabIndex as state can you just check state.active and derive tabIndex based off of that here?

Commit:
refactor keyboardTrap.js and extract arrowUp/Down logic from Menu.js to handleMenuKeyboardTrap

Commit:
Update handleEscape

Commit:
Fix type declaration for ariaExpanded

Commit:
Refactor keyboardTrap to make use of two different functions for previous and next element
-Difficult to follow beforehand when there was one function to get next focusable element

Commit:
Calculate tabIndex based on  getTabIndex
- removes the need to update tabIndex state
